audmath/core/api.py

```diff
@@ -677,6 +677,79 @@ def samples(
     return round(duration * sampling_rate)
 
 
+def similarity(
+        u: typing.Union[typing.Sequence, np.ndarray],
+        v: typing.Union[typing.Sequence, np.ndarray],
+) -> typing.Union[np.floating, np.ndarray]:
+    r"""Cosine similarity between two arrays.
+
+    If the incoming arrays are of size
+    :math:`(1, k)` or :math:`(k,)`,
+    a single similarity value is returned.
+    If the arrays are of size
+    :math:`(n, k)`
+    and :math:`(m, k)`
+    with :math:`n > 1`
+    or :math:`m > 1`,
+    an array of size
+    :math:`(n, m)`
+    with similarities is returned.
+
+    The input arrays can also be provided as
+    :class:`pandas.DataFrame`
+    or :class:`pandas.Series`.
+
+    The cosine similarity is given by
+    :math:`\frac{u \cdot v}{\lVert u\rVert_2 \lVert v\rVert_2}`.
+
+    Args:
+        u: input array
+        v: input array
+
+    Returns:
+        similarity between arrays
+
+    Example:
+        >>> similarity([1, 0], [1, 0])
+        1.0
+        >>> similarity([1, 0], [0, 1])
+        0.0
+        >>> similarity([1, 0], [-1, 0])
+        -1.0
+        >>> similarity([1, 0], [[1, 0], [0, 1]])
+        array([1., 0.])
+        >>> similarity([[1, 0], [0, 1]], [[1, 0], [0, 1]])
+        array([[1., 0.],
+               [0., 1.]])
+        >>> similarity([[1, 0], [0, 1]], [[1, 0], [0, 1], [-1, 0]])
+        array([[ 1.,  0., -1.],
+               [ 0.,  1.,  0.]])
+
+    """
+    def to_numpy(x):
+        if not isinstance(u, np.ndarray):
+            try:
+                # pandas object
+                x = x.to_numpy()
+            except AttributeError:
+                # sequence
+                x = np.array(x)
```

Comment:
I am wondering if this is the nicest way to handle those different types, as I would not expect the most basic input type to be handled in an exception. But it works, therefore no problem?

Reply:
The reason for this slightly ugly code is that we do support `pandas` here without the need of having `pandas` as a dependency for `audmath`, which means we cannot use any code from the `pandas` package.
We just rely on the fact that a user will have `pandas` installed when providing a data frame or series.

If we don't want to use an exception here we could do something like `u.__class__.__name__ == 'DaraFrame' as we do for `Timedelta` inside `audmath.duration_in_seconds()`. But as we need to check for data frame or series here I thought that it might be less error prune to just call `to_numpy()` if it is available.
Also not that `try`-`except` are most of the time faster than doing explicit `if` statements.

But I agree, it doesn't look very nice. If you have a better idea I'm very open to change this ;)

Reply:
I see, thanks for the detailed explanation. I don't have a better idea :)